sync3/caches/global.go

```diff
@@ -101,23 +102,31 @@ func (c *GlobalCache) LoadRooms(ctx context.Context, roomIDs ...string) map[stri
 	return result
 }
 
-// Load all current joined room metadata for the user given. Returns the absolute database position along
-// with the results. TODO: remove with LoadRoomState?
-func (c *GlobalCache) LoadJoinedRooms(ctx context.Context, userID string) (pos int64, joinedRooms map[string]*internal.RoomMetadata, err error) {
+// Load all current joined room metadata for the user given, augmented with the NID of
+// the user's latest join to the room. Returns the absolute database position (the
+// latest event NID across the whole DB) - along with the results.
+// TODO: remove with LoadRoomState?
+func (c *GlobalCache) LoadJoinedRooms(ctx context.Context, userID string) (
+	pos int64, joinedRooms map[string]*internal.RoomMetadata, joinNIDs map[string]int64, err error,
+) {
 	if c.LoadJoinedRoomsOverride != nil {
 		return c.LoadJoinedRoomsOverride(userID)
 	}
 	initialLoadPosition, err := c.store.LatestEventNID()
 	if err != nil {
-		return 0, nil, err
+		return 0, nil, nil, err
 	}
-	joinedRoomIDs, err := c.store.JoinedRoomsAfterPosition(userID, initialLoadPosition)
+	joinedRoomIDs, joinNIDsSlice, err := c.store.JoinedRoomsAfterPosition(userID, initialLoadPosition)
 	if err != nil {
-		return 0, nil, err
+		return 0, nil, nil, err
+	}
+	joinNIDs = make(map[string]int64, len(joinNIDsSlice))
+	for i, nid := range joinNIDsSlice {
+		joinNIDs[joinedRoomIDs[i]] = nid
```

Comment:
Amusing that we immediately zip up the two slices.. can we just make `JoinedRoomsAfterPosition` return `map[string]int64`?

Reply:
b3544a5 attempts this, PTAL